nntrainer/app_context.cpp

```diff
@@ -35,11 +35,22 @@ static void init_global_context_nntrainer(void) __attribute__((constructor));
  */
 static void fini_global_context_nntrainer(void) __attribute__((destructor));
 
-static void init_global_context_nntrainer(void) {}
+static void init_global_context_nntrainer(void) {
+  using OptType = ml::train::OptimizerType;
+
+  auto &ac = AppContext::Global();
+  ac.registerFactory(ml::train::createOptimizer<SGD>, SGD::type, OptType::SGD);
+  ac.registerFactory(ml::train::createOptimizer<Adam>);
+  ac.registerFactory(AppContext::unknownFactory<ml::train::Optimizer>,
+                     "unknown", OptType::UNKNOWN);
+}
 
 static void fini_global_context_nntrainer(void) {}
 
-AppContext &AppContext::Global() { return AppContext::instance; }
+AppContext &AppContext::Global() {
+  static AppContext instance;
```

Comment:
There seems a static init order problem (or might be compiler error...) when I made this to global variable. (http://www.cs.technion.ac.il/users/yechiel/c++-faq/static-init-order.html) which led to a floating point error. (referring to https://stackoverflow.com/questions/13580823/sigfpe-when-accessing-unordered-map)

Changing this to a function local static variable. Since it does not make much of a difference for us.